py-polars/tests/unit/datatypes/test_temporal.py

```diff
@@ -760,6 +765,69 @@ def test_upsample_time_zones(
     assert_frame_equal(result, expected)
 
 
+@pytest.mark.parametrize(
+    ("every", "offset", "fill", "expected"),
+    [
+        (
+            "1i",
+            "0i",
+            "forward",
+            pl.DataFrame(
+                {
+                    "index": [1, 2, 3, 4] + [5, 6, 7],
+                    "groups": ["a"] * 4 + ["b"] * 3,
+                }
+            ),
+        ),
+        (
+            "1i",
+            "1i",
+            "backward",
+            pl.DataFrame(
+                {
+                    "index": [2, 3, 4] + [6, 7],
+                    "groups": ["a"] * 3 + ["b"] * 2,
+                }
+            ),
+        ),
+        (
+            "1i",
+            "1h",
+            "forward",
+            None,
+        ),
+    ],
+)
+def test_upsample_index(
+    every: str,
+    offset: str,
+    fill: FillNullStrategy | None,
+    expected: pl.DataFrame | None,
+) -> None:
+    df = pl.DataFrame(
+        {
+            "index": [1, 2, 4] + [5, 7],
+            "groups": ["a"] * 3 + ["b"] * 2,
+        }
+    ).set_sorted("index")
+
+    if expected is not None:
+        for dtype in [pl.Int32, pl.Int64]:
+            _df = df.with_columns(pl.col("index").cast(dtype))
+            _expected = expected.with_columns(pl.col("index").cast(dtype))
+            result = (
+                _df.upsample(
+                    time_column="index", by="groups", every=every, offset=offset
+                )
+                .fill_null(strategy=fill)
+                .sort(["groups", "index"])
+            )
+            assert_frame_equal(result, _expected)
+    else:
+        with pytest.raises(ComputeError):
+            df.upsample(time_column="index", every=every, offset=offset)
```

Comment:
could you write the test without logic please?

it should just be:
- given: some input
- when: I apply upsample
- then: I check against expected output

the moment you put if/then in tests they become very hard to reason about - split into two tests if necessary

Reply:
@MarcoGorelli Thank you for your review.

I fixed the test.
- split the test into valid and invalid cases
- parameterized datatype and removed for-loop